### PR TITLE
Don't hold KTX files open for longer than transfers require

### PR DESCRIPTION
**Pull request by jherico**
Machines with less CPU RAM seem to suffer under the new KTX system when we hold all the KTX files open all the time, despite memory mapping.  This PR is a first pass at doing more explicit management of the memory mapped files.  Instead of holding the memory mapped files open all the time we instead keep a descriptor of the salient details of the KTX file and have a low-impact process to open a KTX file on demand when we need actual mip data, and then to close it again once we're done.  

## Testing

If you run the current dev build 6209, and go to a domain with a high number of avatars like hifi://dev-chris2.highfidelity.io/100.923,1.6291,27.3687/0,0.998658,0,0.0517909 you can see the private byte count in Task Manager will be very high, on the order of 5-10 GB.  Running with this build the private byte count should stay below 3GB.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: 944745167da259cce2bd208ce4caa8bdc17d67d2 available at: https://gist.github.com/d7685f341c8a1d0c9192549ea0caf44c<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10020-9447451.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10020-9447451.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: f764cd614b90f35e46167f73ee1c162a85e6db36 available at: https://gist.github.com/d7685f341c8a1d0c9192549ea0caf44c<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10020-f764cd6.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10020-f764cd6.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 08db8372c544c9fd0aa6a30157cc8b9f64b4d97b available at: https://gist.github.com/d7685f341c8a1d0c9192549ea0caf44c<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10020-08db837.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10020-08db837.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: e56dcf46e008031dc3801344e0307a9d0086e0bc available at: https://gist.github.com/d7685f341c8a1d0c9192549ea0caf44c<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10020-e56dcf4.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10020-e56dcf4.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 2cf7ba4119dae3fcdfbe0d77854aa6639c751eba available at: https://gist.github.com/d7685f341c8a1d0c9192549ea0caf44c<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10020-2cf7ba4.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10020-2cf7ba4.dmg

File: tests/ktx/src/main.cpp [hunk 1]
```diff
@@ -111,38 +111,40 @@ int main(int argc, char** argv) {
...
-                auto memOffset = memFace - startMemory;
-                auto fileFace = fileImages._faceBytes[face];
-                auto fileOffset = fileFace - startFile;
-                Q_ASSERT(memOffset % 4 == 0);
-                Q_ASSERT(memOffset == fileOffset);
+        auto ktxFile = ktx::KTX::create(std::shared_ptr<storage::Storage>(new storage::FileStorage(TEST_IMAGE_KTX)));
+        {
+            const auto& memStorage = ktxMemory->getStorage();
+            const auto& fileStorage = ktxFile->getStorage();
+            Q_ASSERT(memStorage->size() == fileStorage->size());
```
Comment (samcake) [inline tests/ktx/src/main.cpp]: out of curiosity, why do you use q_assert insert of regular assert?

Comment (jherico) [inline tests/ktx/src/main.cpp]: no particular reason.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 9fc9253eeb5a46557d0254301efad870729d66e9 available at: https://gist.github.com/d7685f341c8a1d0c9192549ea0caf44c<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10020-9fc9253.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10020-9fc9253.dmg

**Comment by OmegaHeron** [comment]
Tested Windows 10 - i5 6500 CPU (8G RAM) - GTX970 GPU

Test setup; Started dev-chris2.highfidelity.io, waited for everything to load and have 100 AVs in play.  Recorded peak mem stat using task manager.

I wonder if I'm doing something wrong here?  Basically - I can't achieve anything like high memory use in 5-10G range cited with Master... it typically runs up to about 2G then down, sometimes as low as 500M.  I can't see any change between Master/this PR.  Is it, perhaps, a difference having an i5 CPU vs i7 could make?  I'm looking at Interface.exe in Task Manager->Details->Memory (private working set) metric.

Desktop mode
Master 6208:  2,166,340K
PR 10020:  Approximately same.

DK2
Same as above.

**Comment by samcake** [comment]
Oups i fucked up
I pressed merged without testing
i then reverted right away
WE are going to reopen that one.
Apologies....

**Pull request merged**